From patchwork Mon Jan 21 12:20:24 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Alexander Popov <alex.popov@linux.com>
X-Patchwork-Id: 10773781
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 32AC2746
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 21 Jan 2019 12:21:08 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 21C1929C10
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 21 Jan 2019 12:21:08 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 155A729C25; Mon, 21 Jan 2019 12:21:08 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id B91E629C16
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 21 Jan 2019 12:21:07 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728129AbfAUMVC (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 21 Jan 2019 07:21:02 -0500
Received: from mail-lf1-f68.google.com ([209.85.167.68]:43928 "EHLO
        mail-lf1-f68.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727926AbfAUMVB (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 21 Jan 2019 07:21:01 -0500
Received: by mail-lf1-f68.google.com with SMTP id u18so15320404lff.10;
        Mon, 21 Jan 2019 04:21:00 -0800 (PST)
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id;
        bh=MnVtObA1fJy4gD+z1gL9LPKcmKRhWMJ7sii8yrEXzQM=;
        b=dGOwpX5RMI2P2azKteXJzfXlLDMvLDfEiY1mnYVgqik7vmhNCy4yEitcVf4IR45k+m
         5Cred5MV/U2jjPgNi2TdO929b5iqzEEBu/DLzor3qoeCs92AqOkcdfishzjcJAip70tu
         M4FXkMPaM9Nhp01hlKlv79JCSm32caf0u5dQrYKJ5ENQ5EHVIMbQ/Je9ubP7HO7W1Nz7
         IqCXRfPqdup1vyVyOLiC6aIbZTeX+BgTvdjuAjcmYWudAKo++eEiPwKCU4oaqq++b7sV
         hXV2lagGprHb+D9OLArz0zLl2H6Hw2lHaKv1W6jTuft/ftYxR1JVSio1CcA/olErVpzn
         rJxw==
X-Gm-Message-State: AJcUukflIFUP36VOdhkm02zRCf6iTo2zbT6zgwBMpp/pKXrq+HJfv39V
        ElN7XJlU4398KEIB9HJ60/Q=
X-Google-Smtp-Source: 
 ALg8bN7DaRXNO4htnXb7bQ9aYJzD0ErtwNDHpFyoNT+R5TOQPCqXKF/fIoRjSqDcmVrUMylc+LF6xg==
X-Received: by 2002:a19:cd50:: with SMTP id
 d77mr17894241lfg.125.1548073259875;
        Mon, 21 Jan 2019 04:20:59 -0800 (PST)
Received: from localhost.localdomain ([213.87.132.221])
        by smtp.gmail.com with ESMTPSA id
 10sm2256354lff.62.2019.01.21.04.20.58
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Mon, 21 Jan 2019 04:20:59 -0800 (PST)
From: Alexander Popov <alex.popov@linux.com>
To: Paolo Bonzini <pbonzini@redhat.com>,
        Radim Krcmar <rkrcmar@redhat.com>,
        Thomas Gleixner <tglx@linutronix.de>,
        Ingo Molnar <mingo@redhat.com>, Borislav Petkov <bp@alien8.de>,
        Nadav Amit <nadav.amit@gmail.com>,
        Andy Lutomirski <luto@kernel.org>,
        Stefan Hajnoczi <stefanha@redhat.com>,
        H Peter Anvin <hpa@zytor.com>, x86@kernel.org,
        kvm@vger.kernel.org, kernel-hardening@lists.openwall.com,
        linux-kernel@vger.kernel.org,
        Alexander Popov <alex.popov@linux.com>
Subject: [PATCH] KVM: x86: Fix single-step debugging
Date: Mon, 21 Jan 2019 15:20:24 +0300
Message-Id: <1548073224-5363-1-git-send-email-alex.popov@linux.com>
X-Mailer: git-send-email 2.7.4
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

The single-step debugging of KVM guests on x86 is broken: if we run
gdb 'stepi' command at the breakpoint when the guest interrupts are
enabled, RIP always jumps to native_apic_mem_write(). Then other
nasty effects follow.

Long investigation showed that the commit c8401dda2f0a00cd25c0 (Jun 7 2017)
introduced the kvm_run.debug corruption: kvm_vcpu_do_singlestep() can
be called without X86_EFLAGS_TF set.

Let's fix it. Please consider that for -stable.
---
 arch/x86/kvm/x86.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index f049ecf..9686068 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -6407,8 +6407,7 @@ int x86_emulate_instruction(struct kvm_vcpu *vcpu,
 		toggle_interruptibility(vcpu, ctxt->interruptibility);
 		vcpu->arch.emulate_regs_need_sync_to_vcpu = false;
 		kvm_rip_write(vcpu, ctxt->eip);
-		if (r == EMULATE_DONE &&
-		    (ctxt->tf || (vcpu->guest_debug & KVM_GUESTDBG_SINGLESTEP)))
+		if (r == EMULATE_DONE && ctxt->tf)
 			kvm_vcpu_do_singlestep(vcpu, &r);
 		if (!ctxt->have_exception ||
 		    exception_type(ctxt->exception.vector) == EXCPT_TRAP)
